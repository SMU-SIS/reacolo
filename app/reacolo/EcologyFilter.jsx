import React from 'react';
import ComponentFilter from './ComponentFilter';

const filters = {
  atLeast(target, context) {
    return target.every(role => context.roleAssignations[role] > 0);
  },
  strict(target, context) {
    return Object.keys(context.roleAssignations).length === target.length
      && filters.atLeast(target, context);
  }
};

const EcologyFilter = (props) => {
  const shouldRender = filters[props.method](props.target, props.context);
  return (
    <ComponentFilter rendered={shouldRender}>
      {React.Children.only(props.children)}
    </ComponentFilter>
  );
};

EcologyFilter.propTypes = {
  target: React.PropTypes.oneOfType([
    React.PropTypes.arrayOf(React.PropTypes.string),
    // TODO:
    // - React.PropTypes.arrayOf(React.PropTypes.arrayOf(React.PropTypes.string))
    // - React.PropTypes.objectOf(React.PropTypes.number)
  ]).isRequired,
  method: React.PropTypes.oneOf(['atLeast', 'strict']),
  children: React.PropTypes.element.isRequired,
  context: React.PropTypes.shape({
    roleAssignations: React.PropTypes.objectOf(React.PropTypes.number).isRequired
  }).isRequired
};

EcologyFilter.defaultProps = {
  method: 'atLeast'
};

export default EcologyFilter;
