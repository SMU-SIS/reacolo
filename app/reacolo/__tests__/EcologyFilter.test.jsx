/* eslint-env jest */

import React from 'react';
import renderer from 'react-test-renderer';
import EcologyFilter from '../EcologyFilter';

test('by default, EcologyFilter renders if all target roles are taken.', () => {
  const target = ['role1', 'role2'];
  const context = {
    roleAssignations: {
      role1: 2,
      role2: 1,
      role3: 1
    }
  };
  const component = renderer.create(
    <div id="container">
      <EcologyFilter target={target} context={context} >
        <div id="child" />
      </EcologyFilter>
    </div>
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('By default, EcologyFilter does not render if there is one target role that is not taken.', () => {
  const target = ['role1', 'role2'];
  const context = {
    roleAssignations: {
      role1: 2,
      role3: 1
    }
  };
  const component = renderer.create(
    <div id="container">
      <EcologyFilter target={target} context={context} >
        <div id="child" />
      </EcologyFilter>
    </div>
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('In strict mode, EcologyFilter renders if all target roles are taken and there is no extraneous roles.', () => {
  const target = ['role1', 'role2'];
  const context = {
    roleAssignations: {
      role1: 2,
      role2: 1
    }
  };
  const component = renderer.create(
    <div id="container">
      <EcologyFilter target={target} context={context} method="strict" >
        <div id="child" />
      </EcologyFilter>
    </div>
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('In strict mode, EcologyFilter does not render if there is a taken role that is not in target.', () => {
  const target = ['role1', 'role2'];
  const context = {
    roleAssignations: {
      role1: 2,
      role2: 1,
      role3: 1
    }
  };
  const component = renderer.create(
    <div id="container">
      <EcologyFilter target={target} context={context} method="strict" >
        <div id="child" />
      </EcologyFilter>
    </div>
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
