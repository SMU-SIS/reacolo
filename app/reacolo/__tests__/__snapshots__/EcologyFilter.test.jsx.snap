exports[`test By default, EcologyFilter does not render if there is one target role that is not taken. 1`] = `
<div
  id="container" />
`;

exports[`test If an array of ecologies is provided, EcologyFilter does not render if none of these ecologies comply. 1`] = `
<div
  id="container" />
`;

exports[`test If an array of ecologies is provided, EcologyFilter renders if more than one of these ecologies comply. 1`] = `
<div
  id="container">
  <div
    id="child" />
</div>
`;

exports[`test If an array of ecologies is provided, EcologyFilter renders if one of these ecologies complies. 1`] = `
<div
  id="container">
  <div
    id="child" />
</div>
`;

exports[`test In strict mode, EcologyFilter does not render if there is a taken role that is not in target. 1`] = `
<div
  id="container" />
`;

exports[`test In strict mode, EcologyFilter renders if all target roles are taken and there is no extraneous roles. 1`] = `
<div
  id="container">
  <div
    id="child" />
</div>
`;

exports[`test by default, EcologyFilter renders if all target roles are taken. 1`] = `
<div
  id="container">
  <div
    id="child" />
</div>
`;
